articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -216,7 +216,27 @@ Furthermore, service code is upgraded one upgrade domain at a time. So, during a
 > While you can modify the schema of a key, you must ensure that your key's hash code and equals algorithms are stable. If you change how either of these algorithms operate, you will not be able to look up the key within the reliable dictionary ever again.
 > .NET Strings can be used as a key but use the string itself as the key--do not use the result of String.GetHashCode as the key.
 
-Alternatively, you can perform what is typically referred to as a two upgrade. With a two-phase upgrade, you upgrade your service from V1 to V2: V2 contains the code that knows how to deal with the new schema change but this code doesn't execute. When the V2 code reads V1 data, it operates on it and writes V1 data. Then, after the upgrade is complete across all upgrade domains, you can somehow signal to the running V2 instances that the upgrade is complete. (One way to signal this is to roll out a configuration upgrade; this is what makes this a two-phase upgrade.) Now, the V2 instances can read V1 data, convert it to V2 data, operate on it, and write it out as V2 data. When other instances read V2 data, they do not need to convert it, they just operate on it, and write out V2 data.
+Alternatively, you can perform a multi-phase upgrade. 
+1. Deploy a new service version that
+    - has both the original version, V1 of the data contracts, and the new version, V2 of the data contracts.
+    - During this phase, the service must continue accessing data in the original, V1 collection using V1 data contracts.
+2. Deploy a new service version where
+    - a new, V2 collection iscreated;
+    - all add, update and delete operations are performed on both V1 and V2 collections in a single transaction;

```

Comment:
What about APIs which combine a read and write? For example GetOrAdd TryUpdate, etc?

If you have

Txn 1:
v1.TryUpdate(key, val1)
v2.TryUpdate(key, val1)

concurrent to Txn 2:
v1.TryUpdate(key, val2)
v2.TryUpdate(key, val2)

Can we end up with:
V1: {key=val1}
V2: {key=val2}

We may need to take the lock on the key in both V1 and V2 before doing either?

https://learn.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-reliable-collections-transactions-locks#isolation-levels

Reply:
Nevermind this is not valid.

Under read repeatable, one txn would wait on acquiring the read lock if the other did the write.

What does seem to matter is the order. It must always be V1 then V2. If there's a mixture, you could have an inconsistency.

Similarly, any reads in the transaction would always have to be from the same collection, I'm guessing V1.

I am still wondering a bit about edge scenarios such as Snapshot transactions concurrent to Read Repeatable ones, but would need to think more...

Reply:
The mixed read+write conditional APIs still concern me. It basically means you'd have

Read V1
Conditional Update V1
Read V2 
Conditional Update V2

When instead  I think you'd want

Read V1
Conditional (on the read) Update V1 + V2

It seems like there could be a messy web of complications with locking?

Reply:
I think the text was already instructing to do what you said we'd want. I've tried to make it more explicit. Does it work now?